salt-memory/salt/memory/allocator_traits.hpp

```diff
@@ -0,0 +1,131 @@
+#pragma once
+#include <salt/meta.hpp>
+
+#include <salt/foundation/fast_terminate.hpp>
+#include <salt/memory/detail/align.hpp>
+
+namespace salt {
+
+namespace detail {
+
+// clang-format off
+template <typename Allocator>
+concept has_allocate_node =
+    requires(Allocator allocator, std::size_t size, std::size_t alignment) {
+        { allocator.allocate_node(size, alignment) } -> std::same_as<void*>;
+    };
+
+template <typename Allocator>
+concept has_deallocate_node =
+    requires(Allocator allocator, void* node, std::size_t size, std::size_t alignment) {
+        { allocator.deallocate_node(node, size, alignment) } -> std::same_as<void*>;
+    };
+
+template <typename Allocator>
+concept has_allocate_array =
+    requires(Allocator allocator, std::size_t count, std::size_t size, std::size_t alignment) {
+        { allocator.allocate_array(count, size, alignment) } -> std::same_as<void*>;
+    };
+
+template <typename Allocator>
+concept has_deallocate_array =
+    requires(Allocator allocator, void* array, std::size_t count, std::size_t size, std::size_t alignment) {
+        { allocator.deallocate_array(array, count, size, alignment) } -> std::same_as<void*>;
+    };
+
+template <typename Allocator>
+concept has_max_node_size =
+    requires(Allocator const allocator) {
+        { allocator.max_node_size() } noexcept -> std::same_as<std::size_t>;
+    };
```

Comment:
The allocator can have `max_node_size()` or `max_size()` methods, this should be taken into account.